File: agents-python/types/risk_scoring_types.py
"""
Risk Scoring Agent - Type Definitions

This module contains all type definitions and data models for the Risk Scoring Agent system.
"""

from dataclasses import dataclass, field
from typing import List, Optional, Dict, Any, Literal
from enum import Enum
from datetime import datetime
from pydantic import BaseModel, Field, validator


class RiskSeverity(str, Enum):
    """Severity levels for risk assessment"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class RiskCategory(str, Enum):
    """Risk categories for scoring"""
    LEGAL = "legal"
    FINANCIAL = "financial"
    COMPLIANCE = "compliance"


class ContractStatus(str, Enum):
    """Contract status after risk assessment"""
    AUTO_APPROVED = "auto_approved"
    REVIEW_RECOMMENDED = "review_recommended"
    MANDATORY_REVIEW = "mandatory_review"
    REJECTED = "rejected"


class ClauseLocation(BaseModel):
    """Location of a clause in the contract"""
    page: Optional[int] = None
    section: Optional[str] = None
    paragraph: Optional[int] = None


class ExtractedClause(BaseModel):
    """Individual extracted clause"""
    clause_type: str
    content: str
    location: ClauseLocation
    importance: Literal["high", "medium", "low"]
    tags: List[str]


class ContractMetadata(BaseModel):
    """Contract metadata"""
    contract_type: str
    jurisdiction: Optional[str] = None
    effective_date: Optional[str] = None
    expiration_date: Optional[str] = None
    parties: List[str]
    contract_value: Optional[float] = None
    currency: Optional[str] = None


class ClauseExtractionOutput(BaseModel):
    """Input from Clause Extractor Agent"""
    contract_id: str
    clauses: List[ExtractedClause]
    metadata: ContractMetadata
    extracted_at: str


class ComplianceViolation(BaseModel):
    """Compliance violation details"""
    standard: str
    requirement: str
    severity: RiskSeverity
    description: str
    recommendation: str
    clause_reference: Optional[str] = None


class ComplianceCheck(BaseModel):
    """Passed compliance checks"""
    standard: str
    requirement: str
    status: Literal["passed", "partial", "not_applicable"]
    clause_reference: Optional[str] = None


class ComplianceCheckOutput(BaseModel):
    """Input from Compliance Check Agent"""
    contract_id: str
    overall_compliance: bool
    violations: List[ComplianceViolation]
    passed_checks: List[ComplianceCheck]
    framework: List[str]
    checked_at: str


class IdentifiedRisk(BaseModel):
    """Individual identified risk"""
    risk_id: str
    category: RiskCategory
    title: str
    description: str
    severity: RiskSeverity
    impact: str
    likelihood: Literal["high", "medium", "low"]
    mitigation: Optional[str] = None
    clause_reference: Optional[str] = None


class RiskAssessmentOutput(BaseModel):
    """Input from Risk Assessment Agent"""
    contract_id: str
    identified_risks: List[IdentifiedRisk]
    overall_risk_level: RiskSeverity
    assessed_at: str


class RiskWeights(BaseModel):
    """Configurable weights for risk scoring"""
    legal: float = 0.33
    financial: float = 0.33
    compliance: float = 0.34

    @validator('*')
    def check_positive(cls, v):
        if v < 0:
            raise ValueError('Weights must be positive')
        return v


class RiskThresholds(BaseModel):
    """Risk thresholds for escalation"""
    low_risk: float = 0.4
    medium_risk: float = 0.7
    high_risk: float = 1.0


class RiskFactor(BaseModel):
    """Contributing risk factor"""
    name: str
    impact: float
    description: str
    source: Literal["clause_extraction", "compliance_check", "risk_assessment"]
    source_reference: Optional[str] = None


class RiskDimensionScore(BaseModel):
    """Individual risk dimension score"""
    category: RiskCategory
    score: float
    severity: RiskSeverity
    factors: List[RiskFactor]
    narrative: str
    confidence: float


class ProcessingMetadata(BaseModel):
    """Processing metadata"""
    agent_version: str
    processing_time_ms: float
    model_used: str
    weights_applied: RiskWeights
    thresholds_applied: RiskThresholds


class RiskScoreOutput(BaseModel):
    """Complete Risk Score output"""
    contract_id: str
    timestamp: str

    # Individual dimension scores
    legal_risk: RiskDimensionScore
    financial_risk: RiskDimensionScore
    compliance_risk: RiskDimensionScore

    # Overall composite score
    composite_score: float
    composite_severity: RiskSeverity

    # Narrative and recommendations
    executive_summary: str
    detailed_narrative: str
    recommendations: List[str]

    # Escalation and status
    status: ContractStatus
    requires_human_review: bool
    escalation_reason: Optional[str] = None

    # Confidence and metadata
    confidence_score: float
    reviewer_notes: Optional[str] = None
    processing_metadata: ProcessingMetadata


class RiskScoringInput(BaseModel):
    """Complete input to Risk Scoring Agent"""
    contract_id: str
    clause_extraction: ClauseExtractionOutput
    compliance_check: ComplianceCheckOutput
    risk_assessment: RiskAssessmentOutput
    custom_weights: Optional[RiskWeights] = None


class ReviewerFeedback(BaseModel):
    """Reviewer feedback for model improvement"""
    reviewer_id: str
    reviewed_at: str
    agreed_with_score: bool
    corrected_score: Optional[Dict[str, float]] = None
    comments: str
    action_taken: Literal["approved", "rejected", "renegotiate", "escalate"]


class RiskScoreRecord(BaseModel):
    """DynamoDB storage record"""
    pk: str  # contract_id
    sk: str  # timestamp
    contract_id: str
    risk_score: RiskScoreOutput
    reviewer_feedback: Optional[ReviewerFeedback] = None
    created_at: str
    updated_at: str
    ttl: Optional[int] = None


class HumanLoopRequest(BaseModel):
    """A2I Human Loop Request"""
    flow_definition_arn: str
    human_loop_name: str
    input_content: Dict[str, Any]
    data_attributes: Optional[Dict[str, List[str]]] = None


class HumanLoopResponse(BaseModel):
    """A2I Human Loop Response"""
    human_loop_arn: str
    human_loop_name: str
    status: Literal["InProgress", "Completed", "Failed", "Stopped"]
    output_destination: Optional[str] = None


class RiskScoringConfig(BaseModel):
    """Configuration for Risk Scoring Agent"""
    weights: RiskWeights = Field(default_factory=RiskWeights)
    thresholds: RiskThresholds = Field(default_factory=RiskThresholds)

    # AWS Configuration
    aws_region: str = "us-east-1"
    dynamodb_table_name: str = "RiskScoringResults"
    s3_bucket_name: str = "contract-risk-reports"

    # Bedrock Configuration
    bedrock_model_id: str = "anthropic.claude-v2"
    bedrock_max_tokens: int = 4096
    bedrock_temperature: float = 0.3

    # A2I Configuration
    a2i_flow_definition_arn: str = ""

    # Escalation Settings
    enable_auto_escalation: bool = True
    enable_reviewer_sampling: bool = True
    sampling_rate: float = 0.1


class ErrorResponse(BaseModel):
    """Error response"""
    error: str
    message: str
    contract_id: Optional[str] = None
    timestamp: str
    details: Optional[Any] = None

File: agents/types/risk-scoring.types.ts
/**
 * Risk Scoring Agent - Type Definitions
 *
 * This file contains all TypeScript interfaces and types for the Risk Scoring Agent system.
 */

/**
 * Severity levels for risk assessment
 */
export enum RiskSeverity {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical'
}

/**
 * Risk categories for scoring
 */
export enum RiskCategory {
  LEGAL = 'legal',
  FINANCIAL = 'financial',
  COMPLIANCE = 'compliance'
}

/**
 * Contract status after risk assessment
 */
export enum ContractStatus {
  AUTO_APPROVED = 'auto_approved',
  REVIEW_RECOMMENDED = 'review_recommended',
  MANDATORY_REVIEW = 'mandatory_review',
  REJECTED = 'rejected'
}

/**
 * Input from Clause Extractor Agent
 */
export interface ClauseExtractionOutput {
  contractId: string;
  clauses: ExtractedClause[];
  metadata: ContractMetadata;
  extractedAt: string;
}

/**
 * Individual extracted clause
 */
export interface ExtractedClause {
  clauseType: string;
  content: string;
  location: {
    page?: number;
    section?: string;
    paragraph?: number;
  };
  importance: 'high' | 'medium' | 'low';
  tags: string[];
}

/**
 * Contract metadata
 */
export interface ContractMetadata {
  contractType: string;
  jurisdiction?: string;
  effectiveDate?: string;
  expirationDate?: string;
  parties: string[];
  contractValue?: number;
  currency?: string;
}

/**
 * Input from Compliance Check Agent
 */
export interface ComplianceCheckOutput {
  contractId: string;
  overallCompliance: boolean;
  violations: ComplianceViolation[];
  passedChecks: ComplianceCheck[];
  framework: string[]; // e.g., ['GDPR', 'HIPAA', 'SOC2']
  checkedAt: string;
}

/**
 * Compliance violation details
 */
export interface ComplianceViolation {
  standard: string; // e.g., 'GDPR', 'HIPAA'
  requirement: string;
  severity: RiskSeverity;
  description: string;
  recommendation: string;
  clauseReference?: string;
}

/**
 * Passed compliance checks
 */
export interface ComplianceCheck {
  standard: string;
  requirement: string;
  status: 'passed' | 'partial' | 'not_applicable';
  clauseReference?: string;
}

/**
 * Input from Risk Assessment Agent
 */
export interface RiskAssessmentOutput {
  contractId: string;
  identifiedRisks: IdentifiedRisk[];
  overallRiskLevel: RiskSeverity;
  assessedAt: string;
}

/**
 * Individual identified risk
 */
export interface IdentifiedRisk {
  riskId: string;
  category: RiskCategory;
  title: string;
  description: string;
  severity: RiskSeverity;
  impact: string;
  likelihood: 'high' | 'medium' | 'low';
  mitigation?: string;
  clauseReference?: string;
}

/**
 * Complete input to Risk Scoring Agent
 */
export interface RiskScoringInput {
  contractId: string;
  clauseExtraction: ClauseExtractionOutput;
  complianceCheck: ComplianceCheckOutput;
  riskAssessment: RiskAssessmentOutput;
  customWeights?: RiskWeights;
}

/**
 * Configurable weights for risk scoring
 */
export interface RiskWeights {
  legal: number;      // Default: 0.33
  financial: number;  // Default: 0.33
  compliance: number; // Default: 0.34
}

/**
 * Individual risk dimension score
 */
export interface RiskDimensionScore {
  category: RiskCategory;
  score: number; // 0.0 to 1.0
  severity: RiskSeverity;
  factors: RiskFactor[];
  narrative: string;
  confidence: number; // 0.0 to 1.0
}

/**
 * Contributing risk factor
 */
export interface RiskFactor {
  name: string;
  impact: number; // 0.0 to 1.0
  description: string;
  source: 'clause_extraction' | 'compliance_check' | 'risk_assessment';
  sourceReference?: string;
}

/**
 * Complete Risk Score output
 */
export interface RiskScoreOutput {
  contractId: string;
  timestamp: string;

  // Individual dimension scores
  legalRisk: RiskDimensionScore;
  financialRisk: RiskDimensionScore;
  complianceRisk: RiskDimensionScore;

  // Overall composite score
  compositeScore: number; // 0.0 to 1.0
  compositeSeverity: RiskSeverity;

  // Narrative and recommendations
  executiveSummary: string;
  detailedNarrative: string;
  recommendations: string[];

  // Escalation and status
  status: ContractStatus;
  requiresHumanReview: boolean;
  escalationReason?: string;

  // Confidence and metadata
  confidenceScore: number; // 0.0 to 1.0
  reviewerNotes?: string;
  processingMetadata: ProcessingMetadata;
}

/**
 * Processing metadata
 */
export interface ProcessingMetadata {
  agentVersion: string;
  processingTimeMs: number;
  modelUsed: string;
  weightsApplied: RiskWeights;
  thresholdsApplied: RiskThresholds;
}

/**
 * Risk thresholds for escalation
 */
export interface RiskThresholds {
  lowRisk: number;      // Default: 0.4
  mediumRisk: number;   // Default: 0.7
  highRisk: number;     // Default: 1.0
}

/**
 * DynamoDB storage record
 */
export interface RiskScoreRecord {
  pk: string; // contractId
  sk: string; // timestamp
  contractId: string;
  riskScore: RiskScoreOutput;
  reviewerFeedback?: ReviewerFeedback;
  createdAt: string;
  updatedAt: string;
  ttl?: number; // Unix timestamp for data expiration
}

/**
 * Reviewer feedback for model improvement
 */
export interface ReviewerFeedback {
  reviewerId: string;
  reviewedAt: string;
  agreedWithScore: boolean;
  correctedScore?: {
    legal?: number;
    financial?: number;
    compliance?: number;
    composite?: number;
  };
  comments: string;
  actionTaken: 'approved' | 'rejected' | 'renegotiate' | 'escalate';
}

/**
 * A2I (Amazon Augmented AI) Human Loop Request
 */
export interface HumanLoopRequest {
  flowDefinitionArn: string;
  humanLoopName: string;
  inputContent: {
    contractId: string;
    riskScore: RiskScoreOutput;
    contractUrl: string;
  };
  dataAttributes?: {
    contentClassifiers: string[];
  };
}

/**
 * A2I Human Loop Response
 */
export interface HumanLoopResponse {
  humanLoopArn: string;
  humanLoopName: string;
  status: 'InProgress' | 'Completed' | 'Failed' | 'Stopped';
  outputDestination?: string;
}

/**
 * Configuration for Risk Scoring Agent
 */
export interface RiskScoringConfig {
  weights: RiskWeights;
  thresholds: RiskThresholds;
  bedrockConfig: {
    modelId: string;
    region: string;
    maxTokens: number;
    temperature: number;
  };
  dynamoDbConfig: {
    tableName: string;
    region: string;
  };
  s3Config: {
    bucketName: string;
    region: string;
  };
  a2iConfig: {
    flowDefinitionArn: string;
    region: string;
  };
  enableAutoEscalation: boolean;
  enableReviewerSampling: boolean;
  samplingRate: number; // 0.0 to 1.0
}

/**
 * Error response
 */
export interface ErrorResponse {
  error: string;
  message: string;
  contractId?: string;
  timestamp: string;
  details?: any;
}
